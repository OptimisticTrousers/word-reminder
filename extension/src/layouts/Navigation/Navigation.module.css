.navigation__top {
  display: flex;
  background-color: var(--top-navigation-bkg-color);
  padding: 0.75rem;
  border-bottom-right-radius: 8px;
  margin-bottom: 0.5rem;
}

.navigation__image,
.navigation__link,
.navigation__item--bottom {
  margin-right: 0.5rem;
}

.navigation__logo {
  display: flex;
  align-items: center;
  margin-right: auto;
}

.navigation__image {
  width: 50px;
  height: 50px;
}

.navigation__heading {
  font-size: 1.2rem;
  line-height: 1.1;
}

.navigation__list {
  display: flex;
  align-items: center;
  list-style: none;
}

.navigation__link {
  padding: 0.5rem;
  display: block;
  text-decoration: none;
  border-radius: 4px;
}

.navigation__link--bottom {
  color: var(--secondary-anchor-text-color);
}

.navigation__link--settings {
  font-size: 1.75rem;
}

.navigation__link--active {
  background-color: var(--active-navigation-bkg-color);
}

.navigation__list--bottom {
  padding: 0.5rem;
  background-color: var(--bottom-navigation-bkg-color);
}

.navigation__item--bottom {
  text-align: center;
  flex: 1;
  border-radius: 4px;
  transition: 0.3s ease;
}

.navigation__link--bottom:hover {
  background-color: var(--tertiary-white-button-bkg-color);
  color: var(--white-button-text-color);
}

.navigation__button:hover {
  opacity: 0.7;
}
