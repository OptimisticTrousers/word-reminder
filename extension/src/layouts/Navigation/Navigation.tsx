import CSSModules from "react-css-modules";
import { Link, useLocation, useNavigate } from "react-router-dom";

import styles from "./Navigation.module.css";
import { useMutation } from "@tanstack/react-query";
import { sessionService } from "../../services/session_service";
import { useNotificationError } from "../../hooks/useNotificationError";

export const Navigation = CSSModules(
  function () {
    const { pathname } = useLocation();
    const navigate = useNavigate();
    const { showNotificationError } = useNotificationError();
    const { isPending, mutate } = useMutation({
      mutationFn: sessionService.logoutUser,
      onSuccess: () => {
        navigate("/login");
      },
      onError: showNotificationError,
    });

    let isUserWordsActive = false;
    let isWordRemindersActive = false;
    let isSettingsActive = false;

    if (pathname === "/userWords") {
      isUserWordsActive = true;
    } else if (pathname === "/wordReminders") {
      isWordRemindersActive = true;
    } else if (pathname === "/settings") {
      isSettingsActive = true;
    }

    function handleLogout() {
      mutate();
    }

    const disabled = isPending;

    return (
      <nav styleName="navigation">
        <Link
          to="/userWords"
          styleName={`navigation__link ${
            isUserWordsActive && "navigation__link--active"
          }`}
          aria-current={isUserWordsActive && "page"}
        >
          User Words
        </Link>
        <Link
          to="/wordReminders"
          styleName={`navigation__link ${
            isWordRemindersActive && "navigation__link--active"
          }`}
          aria-current={isWordRemindersActive && "page"}
        >
          Word Reminders
        </Link>
        <Link
          to="/settings"
          styleName={`navigation__link ${
            isSettingsActive && "navigation__link--active"
          }`}
          aria-current={isSettingsActive && "page"}
        >
          Settings
        </Link>
        <button
          styleName="navigation__button"
          onClick={handleLogout}
          disabled={disabled}
        >
          Log Out
        </button>
      </nav>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
