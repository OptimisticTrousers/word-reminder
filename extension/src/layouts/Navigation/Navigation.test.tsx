import { createRoutesStub, Outlet } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import { Navigation } from "./Navigation";
import userEvent from "@testing-library/user-event";

function UserWords() {
  return <div data-testid="userWords">user words</div>;
}

function WordReminders() {
  return <div data-testid="wordReminders">word reminders</div>;
}

describe("Navigation component", () => {
  it("renders two links to words and word reminders pages respectively", async () => {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: Navigation,
      },
    ]);

    const { asFragment } = render(<Stub initialEntries={["/"]} />);

    const navigation = screen.getByRole("navigation");
    expect(navigation).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("adds the active class to the words page when on the words page", async () => {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return (
            <>
              <Navigation />
              <Outlet />
            </>
          );
        },
        children: [
          {
            path: "/userWords",
            Component: UserWords,
          },
          { path: "/wordReminders", Component: WordReminders },
        ],
      },
    ]);
    const user = userEvent.setup();
    render(<Stub initialEntries={["/wordReminders"]} />);

    const userWordsLink = screen.getByRole("link", {
      name: "User Words",
      current: false,
    });
    await user.click(userWordsLink);

    const userWords = screen.getByTestId("userWords");
    expect(userWordsLink.getAttribute("class")).toContain(
      "navigation__link--active"
    );
    expect(userWords).toBeInTheDocument();
    expect(userWordsLink).toHaveAttribute("aria-current", "page");
  });

  it("adds the active class to the word reminders page when on the word reminders page", async () => {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return (
            <>
              <Navigation />
              <Outlet />
            </>
          );
        },
        children: [
          {
            path: "/userWords",
            Component: UserWords,
          },
          { path: "/wordReminders", Component: WordReminders },
        ],
      },
    ]);
    const user = userEvent.setup();

    render(<Stub initialEntries={["/userWords"]} />);

    const wordRemindersLink = screen.getByRole("link", {
      name: "Word Reminders",
      current: false,
    });
    await user.click(wordRemindersLink);

    const wordReminders = screen.getByTestId("wordReminders");
    expect(wordRemindersLink.getAttribute("class")).toContain(
      "navigation__link--active"
    );
    expect(wordReminders).toBeInTheDocument();
    expect(wordRemindersLink).toHaveAttribute("aria-current", "page");
  });
});
