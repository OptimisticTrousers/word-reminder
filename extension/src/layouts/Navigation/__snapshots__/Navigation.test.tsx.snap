// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation component > renders two links to words and word reminders pages respectively 1`] = `
<DocumentFragment>
  <nav
    class="_navigation_511cb9"
  >
    <a
      aria-current="false"
      class="_navigation__link_511cb9 _false_511cb9"
      data-discover="true"
      href="/userWords"
    >
      User Words
    </a>
    <a
      aria-current="false"
      class="_navigation__link_511cb9 _false_511cb9"
      data-discover="true"
      href="/wordReminders"
    >
      Word Reminders
    </a>
    <a
      aria-current="false"
      class="_navigation__link_511cb9 _false_511cb9"
      data-discover="true"
      href="/settings"
    >
      Settings
    </a>
    <button
      class="_navigation__button_511cb9"
    >
      Log Out
    </button>
  </nav>
</DocumentFragment>
`;
